### PR TITLE
Refactor Stream/Writer; write as C++

### PR DESCRIPTION
**Pull request by binji**
I also took the opportunity to make classes for
ASTWriter/BinaryWriter/BinaryWriterSpec.

### CODE REVIEW

**Comment by binji** [comment]
Hm, probably should have left the other class refactors for another PR, I'll revert those first.

**Comment by binji** [comment]
OK, that's a bit simpler.

File: src/stream.cc [hunk 1]
```diff
@@ -24,140 +24,110 @@
...
-static FileStream s_stdout_stream;
-static FileStream s_stderr_stream;
-
-void init_stream(Stream* stream, Writer* writer, Stream* log_stream) {
-  stream->writer = writer;
-  stream->offset = 0;
-  stream->result = Result::Ok;
-  stream->log_stream = log_stream;
+Stream::Stream(Writer* writer, Stream* log_stream)
+    : writer_(writer),
```
Comment (sbc100) [inline src/stream.cc]: Oh my, the trailing underscore naming convention is so ugly, no?  I don't understand why anyone would choose this...but I guess its our chosen style for chrome?

File: src/stream.cc [hunk 2]
```diff
@@ -24,140 +24,110 @@
...
+MemoryStream::MemoryStream() : Stream(&writer_) {}
+
+FileStream::FileStream(const char* filename)
+    : Stream(&writer_), writer_(filename) {}
+
+FileStream::FileStream(FILE* file) : Stream(&writer_), writer_(file) {}
+
+// static
+std::unique_ptr<FileStream> FileStream::CreateStdout() {
+  return std::unique_ptr<FileStream>(new FileStream(stdout));
```
Comment (sbc100) [inline src/stream.cc]: Isn't 'make_unique' designed to do this kind of thing?

File: src/stream.h [hunk 3]
```diff
@@ -17,82 +17,142 @@
...
+
+class FileStream : public Stream {
+ public:
+  explicit FileStream(const char* filename);
+  explicit FileStream(FILE*);
+
+  FileWriter& writer() { return writer_; }
+
+  static std::unique_ptr<FileStream> CreateStdout();
+  static std::unique_ptr<FileStream> CreateStderr();
```
Comment (sbc100) [inline src/stream.h]: Why not make these functions outside the class?

File: src/tools/wasm2wast.cc [hunk 4]
```diff
@@ -160,17 +156,9 @@ int main(int argc, char** argv) {
...
-        } else {
-          init_file_writer_existing(&file_writer, stdout);
-        }
-
-        if (WABT_SUCCEEDED(result)) {
-          result = write_ast(&file_writer.base, &module);
-          close_file_writer(&file_writer);
-        }
+        std::unique_ptr<FileWriter> writer(s_outfile ? new FileWriter(s_outfile)
+                                                     : new FileWriter(stdout));
```
Comment (sbc100) [inline src/tools/wasm2wast.cc]: make_unique?

Comment (sbc100) [inline src/tools/wasm2wast.cc]: put the conditional inside the constructor arguments?

File: src/tools/wast-desugar.cc [hunk 5]
```diff
@@ -149,17 +149,9 @@ int main(int argc, char** argv) {
...
-      } else {
-        init_file_writer_existing(&file_writer, stdout);
-      }
-
-      if (WABT_SUCCEEDED(result)) {
-        result = write_ast(&file_writer.base, module);
-        close_file_writer(&file_writer);
-      }
+      std::unique_ptr<FileWriter> writer(s_outfile ? new FileWriter(s_outfile)
+                                                   : new FileWriter(stdout));
```
Comment (sbc100) [inline src/tools/wast-desugar.cc]: same

File: src/writer.cc [hunk 6]
```diff
@@ -22,177 +22,123 @@
...
-static Result move_data_in_output_buffer(size_t dst_offset,
-                                         size_t src_offset,
-                                         size_t size,
-                                         void* user_data) {
-  MemoryWriter* writer = static_cast<MemoryWriter*>(user_data);
+Result MemoryWriter::MoveData(size_t dst_offset,
+                              size_t src_offset,
+                              size_t size) {
+  if (size == 0) {
+    return Result::Ok;
```
Comment (sbc100) [inline src/writer.cc]: This check wasn't needed before?

File: src/writer.cc [hunk 7]
```diff
@@ -22,177 +22,123 @@
...
-                                    OutputBuffer* out_buf) {
-  *out_buf = writer->buf;
-  writer->buf.start = nullptr;
-  writer->buf.size = 0;
-  writer->buf.capacity = 0;
+  // TODO(binji): this is pretty cheesy, should come up with a better API.
+  if (!file_) {
+    ERROR("fopen name=\"%s\" failed, errno=%d\n", filename, errno);
+  }
+  should_close_ = file_ != nullptr;
```
Comment (sbc100) [inline src/writer.cc]: should_close_(false) in the init list and then set to true here (or in else block if you prefer)?   since you already check for !file_ above.

File: src/writer.cc [hunk 8]
```diff
@@ -22,177 +22,123 @@
...
 }
 
-void close_mem_writer(MemoryWriter* writer) {
-  destroy_output_buffer(&writer->buf);
+FileWriter::~FileWriter() {
+  if (should_close_) {
+    fclose(file_);
```
Comment (sbc100) [inline src/writer.cc]: Why not just use file_ != nullptr for should_close?

File: CMakeLists.txt [hunk 9]
```diff
@@ -220,6 +220,8 @@ add_library(libwabt STATIC
   ${AST_PARSER_GEN_C}
   src/type-checker.cc
   src/validator.cc
+  src/stream.cc
+  src/writer.cc
```
Comment (binji) [inline CMakeLists.txt]: Oh, just random. I move stuff around when I'm compiling so I get errors earlier. Probably should just alphabetize.

File: src/binary-reader-interpreter.cc [hunk 10]
```diff
@@ -1492,24 +1485,19 @@ Result read_binary_interpreter(InterpreterEnvironment* env,
...
       new DefinedInterpreterModule(istream_offset);
 
-  BinaryReaderInterpreter reader(env, module, istream_offset, error_handler);
-  if (WABT_FAILED(reader.Init())) {
-    delete module;
-    return Result::Error;
-  }
-
   InterpreterEnvironmentMark mark = mark_interpreter_environment(env);
+  BinaryReaderInterpreter reader(env, module, istream_offset, error_handler);
```
Comment (binji) [inline src/binary-reader-interpreter.cc]: This is pretty subtle, probably deserves a comment. The reader takes ownership of the istream, which the mark reads from. When it's below, it reads from a null pointer.

File: src/ast-writer.cc [hunk 11]
```diff
@@ -136,7 +136,7 @@ static void WABT_PRINTF_FORMAT(2, 3)
 }
 
 static void write_putc(Context* ctx, char c) {
-  write_data(&ctx->stream, &c, 1, nullptr);
+  ctx->stream.WriteData(&c, 1, nullptr);
```
Comment (binji) [inline src/ast-writer.cc]: Done.

File: src/binary-reader-logging.cc [hunk 12]
```diff
@@ -67,11 +67,11 @@ void BinaryReaderLogging::WriteIndent() {
   static size_t s_indent_len = sizeof(s_indent) - 1;
```
Comment (binji) [inline src/binary-reader-logging.cc]: Done.

File: CMakeLists.txt (hunk 9 above)
Comment (sbc100) [inline CMakeLists.txt]: Revert his then?

File: src/binary-writer-spec.cc [hunk 13]
```diff
@@ -312,15 +307,9 @@ static void write_raw_module(Context* ctx,
...
-    FileStream stream;
-    Result result = init_file_writer(&stream.writer, filename);
-    if (WABT_SUCCEEDED(result)) {
-      init_stream(&stream.base, &stream.writer.base, nullptr);
-      write_data(&stream.base, raw_module->binary.data, raw_module->binary.size,
-                 "");
-      close_file_writer(&stream.writer);
-    }
-    ctx->result = result;
+    FileStream stream(filename);
```
Comment (binji) [inline src/binary-writer-spec.cc]: Done.

File: src/binary-writer-spec.cc [hunk 14]
```diff
@@ -312,15 +307,9 @@ static void write_raw_module(Context* ctx,
...
+    stream.WriteData(raw_module->binary.data, raw_module->binary.size, "");
```
Comment (binji) [inline src/binary-writer-spec.cc]: Yeah, can just use the default here (nullptr).

File: src/binary-writer.cc [hunk 15]
```diff
@@ -945,16 +939,13 @@ static Result write_module(Context* ctx, const Module* module) {
...
 
 Result write_binary_module(Writer* writer,
                            const Module* module,
                            const WriteBinaryOptions* options) {
-  Context ctx;
-  ctx.options = options;
-  ctx.log_stream = options->log_stream;
-  init_stream(&ctx.stream, writer, ctx.log_stream);
+  Context ctx(writer, options);
   return write_module(&ctx, module);
```
Comment (binji) [inline src/binary-writer.cc]: Yeah, I had a refactor in a earlier commit (making this whole thing into a class), but this makes for a simpler change, I think.

File: src/stream.cc (hunk 2 above)
Comment (binji) [inline src/stream.cc]: Yes, but it's only in C++14. I guess we should have our own version like binaryen does.  I was thinking that would be nicer to do in a separate PR.

File: CMakeLists.txt (hunk 9 above)
Comment (binji) [inline CMakeLists.txt]: Yeah, sorry just going through comments as I fix stuff locally.

File: src/stream.cc (hunk 1 above)
Comment (binji) [inline src/stream.cc]: Yeah, I don't really care either way. But it is better than m_foo. And I like making a distinction between private members and local variables.

File: src/tools/wasm2wast.cc (hunk 4 above)
Comment (binji) [inline src/tools/wasm2wast.cc]: Can't, they're different constructors.

File: src/writer.cc (hunk 6 above)
Comment (binji) [inline src/writer.cc]: Hm, not sure why I added it.

File: src/writer.cc (hunk 7 above)
Comment (binji) [inline src/writer.cc]: Done.

File: src/writer.cc (hunk 8 above)
Comment (binji) [inline src/writer.cc]: since we don't want to close the file when it's passed in (stdout, for example)

File: src/stream.h (hunk 3 above)
Comment (binji) [inline src/stream.h]: Could do, I guess. This way the names are scoped, which is nice. Also, seems reasonable to have constructor functions be defined as static members.

File: src/tools/wasm2wast.cc (hunk 4 above)
Comment (binji) [inline src/tools/wasm2wast.cc]: Changed it to just call write_ast in each branch, nicer this way.

File: src/tools/wasm2wast.cc [hunk 16]
```diff
@@ -156,9 +156,13 @@ int main(int argc, char** argv) {
...
       if (WABT_SUCCEEDED(result)) {
-        std::unique_ptr<FileWriter> writer(s_outfile ? new FileWriter(s_outfile)
-                                                     : new FileWriter(stdout));
-        result = write_ast(writer.get(), &module);
+        if (s_outfile) {
+          FileWriter writer(s_outfile);
+          result = write_ast(&writer, &module);
+        } else {
+          FileWriter writer(stdout);
+          result = write_ast(&writer, &module);
```
Comment (sbc100) [inline src/tools/wasm2wast.cc]: What about FileWriter writer(s_outfile ? s_outfile : stdout) ?

Comment (binji) [inline src/tools/wasm2wast.cc]: That's what I meant before, it's not allowed because they are different constructors. The ternary operator happens first which wants to coerce `const char*` and `FILE*` to some common type but can't.

Comment (binji) [inline src/tools/wasm2wast.cc]: To make this nicer, I added a move constructor to `FileWriter`, so now it's:
```
FileWriter writer(s_outfile ? FileWriter(s_outfile) : FileWriter(stdout));
```

**Comment by binji** [comment]
Annoyingly, still don't have appveyor + travis builds. I've tested locally and everything seems OK. Guess I'll resolve any issues that come up later.

**Comment by dschuff** [comment]
The wabt github services seem to be up again (e.g. https://travis-ci.org/WebAssembly/wabt/builds/222855450). Still can't get Binaryen to work though...

**Comment by binji** [comment]
Looks like it's working to me: https://travis-ci.org/WebAssembly/binaryen

**Comment by dschuff** [comment]
Yeah I think it just came back online in the last half hour or so.
